fastar/update_rules.py

```diff
@@ -21,66 +22,147 @@ def take(arr, idxs):
     assert ans.shape == vals.shape
     return arr + take_vjp(vals)[0]
 
-def _init_out(func, *args, **params):
+
+def _init_output(func, *args, init_value=0, **params):
     # TODO: generalize to jaxvals
     args = map(make_shaped_array, args)
     abstract_out = func.abstract_eval(*args, **params)
-    outval = lax.full(abstract_out.shape, 0, abstract_out.dtype)
+    outval = lax.full(abstract_out.shape, init_value, abstract_out.dtype)
     outmask = onp.zeros(abstract_out.shape, bool)
     return outval, outmask
 
+
+def sliceableop_update(func, old_out, *args, output_mask, input_slices_from_output_slice,
+                       init_output=_init_output, sliced_output_from_sliced_inputs=None,
+                       init_value=0,
+                       **params):
```

Comment:
I’m still sceptical about whether this is a good idea. It reduces the number of lines of code, but in my opinion it also reduces readability/clarity of what’s going on by adding a layer of indirection. What do you think?

Reply:
I find it improves clarity by structuring and naming most relevant parts. sliceableop has a substantial amount of responsibility, e. g. mask initialization, diffing and corresponding slicing, which would be duplicated for each op class. It would further be duplicated for ops we might want to implement in the future, e. g. convs and matmul.

Reply:
I agree it would be good to get rid of the duplicated output initialization, which is going to be the same for _every_ op, i.e. this kind of line:
```
outval, old_outmask = old_out[0] if old_out else _init_out(func, a)
```
However I think we could have a lightweight function (maybe a decorator), that only does that.

Reply:
Maybe we could have another small, separate, helper function which runs the for loop over slices, which I think we also need to do for every op. I.e. two small helper functions instead of one big one.

Reply:
The gain is limited at the moment though since we're covering so few ops.

Reply:
I was able to simplify `sliceableop_update`:

- As you suggested, I moved `_init_out` into the interpreter (replacing `None` in `firstpass`). This saves us from calling the same helper function for every op. 
- Additionally, I made the now quite different case of `pad` independent (see above), saving us 3 arguments.

Since that made it much more readable (could be considered a helper function) and it has six use cases with 2+ more expected in the future, I think it might be a more useful abstraction now. What do you think?